pkg/replicator/scheme.go

```diff
@@ -0,0 +1,389 @@
+
+			return nil
+		}
+	}
+
+	if err := rs.fromBkt.Iter(ctx, chunksDir, func(objectName string) error {
+		err := rs.ensureObjectReplicated(ctx, objectName)
+		if err != nil {
+			return fmt.Errorf("replicate object %v: %w", objectName, err)
+		}
+
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	if err := rs.ensureObjectReplicated(ctx, indexFile); err != nil {
+		return fmt.Errorf("replicate index file: %w", err)
+	}
+
+	level.Debug(rs.logger).Log("msg", "replicating meta file", "object", metaFile)
+
+	if err := rs.toBkt.Upload(ctx, metaFile, bytes.NewReader(originMetaFileContent)); err != nil {
+		return fmt.Errorf("upload meta file: %w", err)
+	}
+
+	rs.metrics.blocksReplicated.Inc()
+
+	return nil
+}
+
+// ensureBlockIsReplicated ensures that an object present in the origin bucket
+// is present in the target bucket.
+func (rs *replicationScheme) ensureObjectReplicated(ctx context.Context, objectName string) error {
+	level.Debug(rs.logger).Log("msg", "ensuring object is replicated", "object", objectName)
+
+	exists, err := rs.toBkt.Exists(ctx, objectName)
+	if err != nil {
+		return fmt.Errorf("check if %v exists in target bucket: %w", objectName, err)
+	}
+
+	// skip if already exists.
+	if exists {
+		level.Debug(rs.logger).Log("msg", "skipping object as already replicated", "object", objectName)
+		return nil
+	}
+
+	level.Debug(rs.logger).Log("msg", "object not present in target bucket, replicating", "object", objectName)
+
+	r, err := rs.fromBkt.Get(ctx, objectName)
+	if err != nil {
+		return fmt.Errorf("get %v from origin bucket: %w", objectName, err)
+	}
+
+	defer r.Close()
+
+	if err = rs.toBkt.Upload(ctx, objectName, r); err != nil {
+		return fmt.Errorf("upload %v to target bucket: %w", objectName, err)
+	}
+
+	level.Info(rs.logger).Log("msg", "object replicated", "object", objectName)
+	rs.metrics.objectsReplicated.Inc()
+
+	return nil
+}
+
+// loadMeta loads the meta.json from the origin bucket and returns the meta
+// struct as well as if failed, whether the failure was due to the meta.json
+// not being present or partial. The distinction is important, as if missing or
+// partial, this is just a temporary failure, as the block is still being
+// uploaded to the origin bucket.
+func loadMeta(ctx context.Context, rs *replicationScheme, id ulid.ULID) (*metadata.Meta, bool, error) {
+	fetcher, err := thanosblock.NewMetaFetcher(rs.logger, 32, rs.fromBkt, "", rs.reg)
+	if err != nil {
+		return nil, false, fmt.Errorf("create meta fetcher with buecket %v: %w", rs.fromBkt, err)
+	}
+
+	metas, _, err := fetcher.Fetch(ctx)
+
+	if err != nil {
+		switch errors.Cause(err) {
+		default:
+			return nil, false, fmt.Errorf("fetch meta: %w", err)
+		case thanosblock.ErrorSyncMetaNotFound:
+			return nil, true, fmt.Errorf("fetch meta: %w", err)
+		}
+	}
+
+	m, ok := metas[id]
+	if !ok {
+		return nil, true, fmt.Errorf("fetch meta: %w", err)
+	}
+
+	if metas[id].Version != metadata.MetaVersion1 {
```

Comment:
Isn't this already checked in `loadMeta` inside of `MetaFetcher.loadMeta`? :P

Reply:
Oh, i see, thanks!